power_grid_model_c/power_grid_model_cpp/include/power_grid_model_cpp/dataset.hpp

```diff
@@ -0,0 +1,175 @@
+// SPDX-FileCopyrightText: Contributors to the Power Grid Model project <powergridmodel@lfenergy.org>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#pragma once
+#ifndef POWER_GRID_MODEL_CPP_DATASET_HPP
+#define POWER_GRID_MODEL_CPP_DATASET_HPP
+
+#include <memory>
+
+#include "dataset.h"
+#include "handle.hpp"
+
+namespace power_grid_model_cpp {
+class Dataset {
+public:
+    power_grid_model_cpp::Handle handle;
+
+    virtual ~Dataset() = default;
+
+    static char const* info_name(PGM_Handle* provided_handle, PGM_DatasetInfo const* info) {
+        return PGM_dataset_info_name(provided_handle, info);
+    }
```

Comment:
This comment is on an outdated version. However I will still ask. My idea of having the `static` or "procedural" version of each member function was to have one that still functions as in the C-API. Hence, shouldn't all the `static` member functions just expect a `Handle` to be passed as an argument (as in giving the user the responsibility of handling that?). Or else, if I want the `static` functions to contain their own handle, wouldn't I have to instantiate `Handle` on each call, which would generate a big overhead?

Reply:
no because: yes, we want it to function as the C API functions, but at the same time, we want to hide away anything that is irrelevant to the user, like the handle. free functions don't mean we still have to provide everything

Reply:
I wish us to have some further discussion on this. For example, in `meta_data.hpp` the free functions won't expect any parameter that has access to a `Handle`. In this case, I believe a singleton approach as you suggested initially would be the way to go, something like
```
static Handle& getHandle() {
    static Handle handle; 
    return handle;
}
```
And to my knowledge, `static` is thread safe from c++11, hence this should not be an issue if the user wants to run its own threads. But maybe I am missing something here.

Moreover, a similar thing happens to DatasetMutable. Maybe there is a workaround I am not seeing, but I am not aware of it if we want to remove the `Handle` as an argument from the free functions.

Reply:
In the case in which a `Handle` could not be provided (only in `meta_data.hpp`), all member functions were made `static` and a `Handle` was individually instantiated on each call. Since these set of functions are not expected to be called excessively if used correctly, the overhead is not significant as discussed.